Display site-wide announcements

vaidehijoshi opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This feature allows active announcement broadcasts to be displayed, site-wide.

Specifically, this PR does the following: 
- [x] Adds a `broadcast:` key to our `AsyncInfoController`'s `base_data`, regardless of whether a user is logged in or not. If a broadcast is not present, passes along a `nil` value.
- [x] Initializes the active broadcast (if present) when rendering _any_ page.
- [x] Allows admins to preview a broadcast's `processed_html` when editing the broadcast on `/internal/broadcasts/:id`.
- [x] Adds specs around serializing and rendering an active broadcast (request + system-level specs).
- [x] Add some temporary styling to the broadcast banner such that it is `fixed` to the top of the page, below the nav. I actually don't really like how this looks, but it ensures that the broadcast renders with the highest `z-index` on an article page. **HOWEVER, once https://github.com/thepracticaldev/dev.to/pull/8234 is merged, the default style on a broadcast will Just Work™️, and we can remove this styling (you'll notice I added a TODO for this).**

The next step in this feature will be https://github.com/thepracticaldev/dev.to/issues/8091, which will allow the user to dismiss this active announcement. But, since this feature isn't being used yet in production and it doesn't affect any users just yet, I'll be handling that in a subsequent PR.

## Related Tickets & Documents
Closes https://github.com/thepracticaldev/dev.to/issues/8090.

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
**TO QA THIS PR:**
1. Go to `/internal/broadcasts`, and create/update a broadcast with some html/inline css. Once you save it, you can preview the broadcast and update it to be active:
<img width="1077" alt="Screen_Shot_2020-06-04_at_4_39_31_PM" src="https://user-images.githubusercontent.com/6921610/83911342-f33ee980-a720-11ea-97df-cae17b608954.png">
2. Go to any page on the site. You should see the broadcast displayed:
<img width="1520" alt="Screen Shot 2020-06-04 at 4 39 48 PM" src="https://user-images.githubusercontent.com/6921610/83911119-980cf700-a720-11ea-9ab7-c6650b7dc3bc.png">
3. Try setting that same broadcast to "inactive" from the internal page. You should no longer see the broadcast displayed anywhere on the site!

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## Are there any post deployment tasks we need to perform?
Nope!

## What gif best describes this PR or how it makes you feel?

![Seth Meyers has an announcement gif](https://media2.giphy.com/media/3o6ZteOz7Uz6bE4l6U/giphy.webp?cid=5a38a5a2ff17f22b5534bfecd9c831f8e3bd581d9903a464&rid=giphy.webp)

vaidehijoshi commented on app/assets/stylesheets/scaffolds.scss:
```diff
@@ -242,6 +242,14 @@ body.hidden-shell {
   z-index: -1;
 }
 
+.broadcast-wrapper {
+  // TODO: [@thepracticaldev/delightful]: Remove these styles once
+  // https://github.com/thepracticaldev/dev.to/pull/8234 is merged.
+  z-index: 101; // Must be higher than the z-index of the sticky-nav.
+  position: fixed;
+  width: 100%;
+}
```
As mentioned in the PR description, these styles can be removed once @ludwiczakpawel's PR, https://github.com/thepracticaldev/dev.to/pull/8234, is merged.

I've tried this out against his changes and we shouldn't need these styles since the default `<div>` styling will Just Work™️. Without these styles, the broadcast looks like this on an article page: 
<img width="1655" alt="Screen Shot 2020-06-04 at 4 56 07 PM" src="https://user-images.githubusercontent.com/6921610/83911705-8d9f2d00-a721-11ea-9c2a-09d6d42a6d0b.png">

So that's why I'm adding this fixed positioning. But with @ludwiczakpawel's PR changes, it will look like this, so I think we'll be able to remove most of this styling:
<img width="1349" alt="Screen Shot 2020-06-05 at 8 43 17 AM" src="https://user-images.githubusercontent.com/6921610/83911721-94c63b00-a721-11ea-9dc3-da3064e26c3a.png">

I'm

vaidehijoshi replied:
Addressed! Added styles to handle the two different header configs (totally forgot that we had that, good catch on pointing that out, @ludwiczakpawel!):

_Sticky:_
<img width="1475" alt="Screen Shot 2020-06-08 at 9 53 18 AM" src="https://user-images.githubusercontent.com/6921610/84058367-009ce380-a96e-11ea-87a2-6e36c994fef6.png">

_Static:_
<img width="1475" alt="Screen Shot 2020-06-08 at 9 53 09 AM" src="https://user-images.githubusercontent.com/6921610/84058384-05619780-a96e-11ea-992d-a90e17ce28fc.png">

vaidehijoshi commented on spec/system/homepage/user_visits_homepage_spec.rb:
```diff
@@ -29,8 +30,28 @@
+      let!(:broadcast) { create(:announcement_broadcast) }
+
+      it "renders the broadcast if active", js: true do
+        get "/async_info/base_data" # Explicitly ensure broadcast data is loaded before doing any checks
```
I don't love that I have to do this, but it seems necessary since, afaik, the request to `async_info/base_data` doesn't happen by default. If someone has better ideas, open to hearing them!

vaidehijoshi commented on app/views/internal/broadcasts/_form.html.erb:
```diff
@@ -14,3 +14,12 @@
+<% if @broadcast.processed_html %>
+  <div>
+    </p><strong>Preview</strong></p>
+    <%= @broadcast.processed_html.html_safe %>
```
Hm, looks like brakeman is complaining that this is an `Unescaped model attribute`? This is a new one for me, but from the source code it looks like brake man [does indeed think](https://github.com/presidentbeef/brakeman/blob/c9fc366db717f759c064371a0321cccc88c2affb/test/apps/rails4/app/views/another/html_safe_is_not.html.erb) that this is unsafe.

Not sure what is the best way to get around this, since we _do_ want to be able to let admins preview here, and this html is a value that comes _directly from the database_ and not a cookie/param. Any thoughts? 🤔

juliannatetreault replied:
Huh... 🤔 Maybe something like this could be useful?: https://stackoverflow.com/questions/39099489/how-to-fix-cross-site-scripting-security-warning-in-rails-when-using-html-safe

vaidehijoshi replied:
@juliannatetreault Great find!! Trying it out with `sanitize`.

Sadly, it seems I need to explicitly pass in the sanitized attributes (according to [the docs](https://api.rubyonrails.org/classes/ActionView/Helpers/SanitizeHelper.html#method-i-sanitize_css) `style` and css should be sanitized by default), but I think that might be because of the way sanitize works with nested elements. In any case, I think passing in explicitly is safer.

Thanks again sending that method my way! 👍

ludwiczakpawel commented on app/views/shell/_top.html.erb:
```diff
@@ -58,6 +58,7 @@
           <%= yield(:audio) %>
         </div>
         <%= render "layouts/top_bar" %>
+        <div id="active-broadcast" class="broadcast-wrapper"></div>
```
Do we want this empty DIV to always be there even if there's no announcement visible? Is there any way to avoid that?

ludwiczakpawel replied:
Also, we do have something like `crayons-banner` (not documented so it's absolutely fair you weren't aware of that :D) that basically supposed to be this kind of a banner... We do have several variations:

```
<div class="crayons-banner">...</div>
```
![image](https://user-images.githubusercontent.com/108287/84005078-d7c81e80-a96c-11ea-8301-f649ceb0fca8.png)


```
<div class="crayons-banner crayons-banner--brand">...</div>
```
![image](https://user-images.githubusercontent.com/108287/84005174-fc23fb00-a96c-11ea-82cb-02bc3b174958.png)

```
<div class="crayons-banner crayons-banner--success">...</div>
```
![image](https://user-images.githubusercontent.com/108287/84005320-3d1c0f80-a96d-11ea-9f00-0c8a2c9897bd.png)

```
<div class="crayons-banner crayons-banner--warning">...</div>
```
![image](https://user-images.githubusercontent.com/108287/84005355-4dcc8580-a96d-11ea-8e54-beb69f5e8f45.png)

```
<div class="crayons-banner crayons-banner--error">...</div>
```
![image](https://user-images.githubusercontent.com/108287/84005406-62108280-a96d-11ea-87b3-a7df155773f7.png)



We could obviously add additional ones and/or expend that component if needed.

vaidehijoshi replied:
@ludwiczakpawel:
> Do we want this empty DIV to always be there even if there's no announcement visible? Is there any way to avoid that?

The only point where we know whether an announcement exists or not is when we fetch it [and assign it from the backend](https://github.com/thepracticaldev/dev.to/pull/8301/files#diff-c9d5cb67f974ae68d44442f67dea97f2R41). When we initialize the page, we check to see if any data [was assigned on fetch](https://github.com/thepracticaldev/dev.to/pull/8301/files#diff-a99b7a23c274fa6b57e161e372ad2e28R24), and that's when we do something with the `<div id="active-broadcast" class="broadcast-wrapper"></div>`. The only other option here is to actually insert the div _into_ the body using `append()`, but I personally find that more confusing and a vector for potential UI bugs: we'd have to know _where_ in the body to append it, and we'd be relying on no other elements in the body being shifted around or accidentally added/removed, otherwise the `<div>` could render in a completely incorrect location. Also, it's worth mentioning that the `<div id="message-notice"></div>` element already exists (and is empty except for when used for chat channels), and it would be odd to do something entirely different to that, I think. 


> Also, we do have something like crayons-banner (not documented ...

Yeah, I was chatting with @lisasy about this! Since our `/internal` pages don't have access to crayons, the "preview" on `/internal/broadcasts/:id/edit` does not render any crayons variables. I can definitely add it, but I don't think that this blocks the PR, and would prefer to do it in a separate PR to avoid this one from getting big :)

ludwiczakpawel replied:
Thanks, that was very helpful explanation :) Even though I don't like empty containers 🙈

vaidehijoshi commented on app/controllers/async_info_controller.rb:
```diff
@@ -37,6 +39,16 @@ def shell_version
+    {
+      title: broadcast&.title,
+      html: broadcast&.processed_html
+    }.to_json
```
@rhymes This is a good point, and I have no idea, honestly. tbqh, I was following [what was in the file](https://github.com/thepracticaldev/dev.to/blob/38634e5316140d82c6872673ac3c317034afce82/app/controllers/async_info_controller.rb#L27), but according to [the rails docs](https://guides.rubyonrails.org/layouts_and_rendering.html#rendering-json):
> You don't need to call `to_json` on the object that you want to render. If you use the `:json` option, render will automatically call `to_json` for you.

It looks like we also use it [here](https://github.com/thepracticaldev/dev.to/blob/38634e5316140d82c6872673ac3c317034afce82/app/controllers/async_info_controller.rb#L37) and [here](https://github.com/thepracticaldev/dev.to/blob/38634e5316140d82c6872673ac3c317034afce82/app/controllers/async_info_controller.rb#L47) for some reason.

I kinda don't want to touch this in the context of this PR but it seems like we should fix it in a separate PR and test it to ensure that there actually _isn't_ a reason we're explicitly doing this.

vaidehijoshi commented on app/assets/javascripts/initializers/initializeBroadcast.js:
```diff
@@ -0,0 +1,30 @@
+ * @function initializeBroadcast
+ */
+function initializeBroadcast() {
+  if (!broadcastData()) {
+    return;
+  }
+  const { html = null } = broadcastData();
```
LOL whoops, I was so focused on fixing my borked tests that I totally didn't even see myself doing this, haha. 🙈

vaidehijoshi commented:
@rhymes I have addressed your comments, and included some default styling as you requested. As I mentioned to @ludwiczakpawel, I'm going to add [more default crayon style options](https://github.com/thepracticaldev/dev.to/pull/8301#discussion_r436826417) in a separate PR :)

@ludwiczakpawel I have addressed your questions and fixed the styling with the default navbar config view!

@joshpuetz The bug you noticed was actually pre-existing, and was just because of the way we were _not_ specifying a default value in the select options in the form. I've gone ahead and fixed that 👍
